File: javascript/robot-simulator/robot-simulator.js
const bearings = [];
bearings[0] = 'east';
bearings[90] = 'north';
bearings[180] = 'west';
bearings[270] = 'south';

const instructions = {
  A: 'advance',
  R: 'turnRight',
  L: 'turnLeft',
};

export class InvalidInputError extends Error {}

export class Robot {
  #coordinates;
  #bearing;

  constructor() {
    this.#coordinates = [0, 0];
    this.#bearing = 90;
  }

  get bearing() { return bearings[this.#bearing]; }

  get coordinates() { return this.#coordinates; }

  at(x, y) {
    this.#coordinates = [x, y];
    return this;
  }

  orient(bearing) {
    if (!bearings.includes(bearing)) throw new InvalidInputError(`invalid orientation ${bearing}`);
    this.#bearing = bearings.indexOf(bearing);
    return this;
  }

  place({ x, y, direction }) {
    return this.at(x, y).orient(direction);
  }

  turn(amount) {
    this.#bearing = (this.#bearing + amount + 360) % 360;
    return this;
  }

  turnLeft() { this.turn(+90); }

  turnRight() { this.turn(-90); }

  advance() {
    const rad = this.#bearing * 2 * Math.PI / 360;
    const [x, y] = this.#coordinates;
    const [dx, dy] = [Math.round(Math.cos(rad)), Math.round(Math.sin(rad))];
    return this.at(x + dx, y + dy);
  }

  static instructions(script) {
    return Array.from(script).map((i) => {
      if (!instructions[i]) throw new InvalidInputError(`no such instruction ${i}`);
      return instructions[i];
    });
  }

  evaluate(script) {
    Robot.instructions(script).forEach(i => this[i]());
  }
}

File: ruby/robot-simulator/robot_simulator.rb
# class comment
class Robot
  # this particular ordering used by #advance
  BEARINGS = %i[east north west south].freeze

  attr_reader :bearing, :coordinates

  def at(*coords)
    @coordinates = coords
  end

  def orient(bearing)
    raise ArgumentError unless BEARINGS.include? bearing
    @bearing = bearing
  end

  def turn_right
    turn(-1)
  end

  def turn_left
    turn(+1)
  end

  def turn(direction)
    idx = BEARINGS.index bearing
    new_idx = (idx + direction) % BEARINGS.length
    orient BEARINGS[new_idx]
  end

  def advance
    rad = (BEARINGS.index(bearing) * 90) * (2 * Math::PI / 360)
    x, y = coordinates
    dx = Math.cos(rad).to_i
    dy = Math.sin(rad).to_i
    at x + dx, y + dy
  end
end

# class comment
class Simulator
  INSTRUCTIONS = { 'L' => :turn_left, 'R' => :turn_right, 'A' => :advance }.freeze

  def instructions(script)
    instructions = script.chars
    raise ArgumentError unless instructions.all? { |i| INSTRUCTIONS.key? i }

    instructions.map { |i| INSTRUCTIONS[i] }
  end

  def place(robot, x: 0, y: 0, direction: :north)
    robot.at x, y
    robot.orient direction
  end

  def evaluate(robot, script)
    instructions(script).each do |i|
      robot.send(i)
    end
  end
end
